src/main/client_context_state.cpp

```diff
@@ -0,0 +1,43 @@
+#include "duckdb/main/client_context_state.hpp"
+
+namespace duckdb {
+
+RegisteredStateIterator::RegisteredStateIterator(RegisteredStateManager &manager, iterator_type iterator_p,
+                                                 idx_t state_version)
+    : manager_ref(manager), iterator(std::move(iterator_p)), state_version(state_version) {
+}
+RegisteredStateIterator &RegisteredStateIterator::operator++() {
+	auto &manager = manager_ref.get();
+
+	lock_guard<mutex> lock(manager.lock);
+	if (manager.state_version != state_version) {
+		throw InternalException("RegisteredStateIterator has been invalidated during iteration");
```

Comment:
A perhaps simpler alternative would be to create a copy of the internal map to iterate over. This means the iterator would provide snapshot semantics (which is fine I think) and allow concurrent modification. 

While the current places in the DuckDB code that are iterating over the registered states seem to be ones where concurrent modification is unexpected, I'm not sure if this limitation is harmful for extensions going forward.

Reply:
I've reworked it to use an `Iterate` method instead that holds the lock while iterating. I would prefer not to make a copy since we iterate quite frequently (and should rarely change the set).

Reply:
We're doing some complex stuff under some of these hooks (e.g. `OnFinalizePrepare`), and would have to be careful not to call `GetOrCreate()` or any other methods there to avoid deadlocking.

> I would prefer not to make a copy since we iterate quite frequently (and should rarely change the set).

One more option here would be to have a CopyOnWriteArrayList approach, i.e. create a new internal map when an element is added, and share the current immutable map with iterators.

Reply:
The way I see the registered states is that there should only be (up to) a single registered state per extension. The `ObjectCache` can be used for caching purposes (the Azure extension currently violates this - that should be fixed separately).

Could you expand on your use case for why you would need to have multiple registered states?

Reply:
Or I guess the worry is that you might call into `httpfs`, which then lazily instantiates its state, creating a deadlock?

I can see that yeah. I'll try to fix it

Reply:
I'm less worried about having multiple registered states. Extensions will typically look up their per-connection extension state via the new `GetOrCreate()` method, retrieved from the ClientContext& that is typically available in many methods. Any ClientContextState methods that are now called as part of the iteration (e.g. QueryBegin, QueryEnd, ... , OnFinalizePrepare, OnExecutePrepared) need to be careful not to retrieve the same extension state again via the ClientContext& that is passed to the method (as that deadlocks), but need to explicitly pass the extension state down to all methods used.

Reply:
I've moved to copying the states after all for simplicity's sake